### PR TITLE
Semi-transparent background for nametags

### PR DESCRIPTION
**Pull request by Zughy**
Sometimes reading names is not that easy: we thought about adding a semi-transparent background to increase contrast. This could work very well paired with #10013 . Also, for testing we set `font_shadow` to false but we did not touch it in the PR

nametags_nobg
nametags

## To do

This PR is Ready for Review.

## How to test
Spawn a few entities and set a name

### CODE REVIEW

**Comment by MoNTE48** [comment]
Should be settable!

**Comment by Zughy** [comment]
@MoNTE48 can you name an use-case? Because the idea is to get rid of the shadow behind the font completely (so I'm/we're planning someday to change the chat too), thus names won't be readable at all without a background. As in the first screenshot.

Like, if only it had been the other way around, would you had proposed it? I mean, to feature nametags without a background

**Comment by rubenwardy** [comment]
I think this is great as an accessibility option, but I'd prefer to have the more subtle names by default

**Comment by MoNTE48** [comment]
@Zughy, if you want to add something to Minetest, make it an option. Otherwise, it will be accepted... never.

**Comment by Zughy** [comment]
@MoNTE48 If true, that's bad. Right now names are simply not readable if they're not pure white. The same goes for chat messages, so it's a pretty big limitation. AT LEAST if it's an option AND #10013 is going to be merged, I'd say to use the background (and the aforementioned PR's behaviour) as default. Also, where should this option be toggled? Server side via minetest.conf? Client side via settings?

I'm actually going to open an issue about it, to keep here clean

**Comment by MoNTE48** [comment]
@Zughy, object (player, entity) parameter. So that you can customize this for each player and object.
Then, I am sure, they will quickly accept it, players will like it, but core devs will never turn it on by default. 🧐
Btw, I press 👍, I really like this idea and I completely agree with what you said above.

**Comment by Zughy** [comment]
@MoNTE48 I think the default option should be the "generally more suited" option. If names are not that readable by default, it's not that well suited and new players will judge it as negative. Show those pictures to five different people who never played Minetest and if they tell you they do prefer the former (the one without bg when it comes to readability), then I'll be proved wrong. Actually, I invite anyone to do that.

If Minetest is "Minecraft-like", then let's cut the chase and adopt what works on Minecraft and let's improve it (as the PR mentioned in the first post, a feature that Minecraft doesn't have).

**Comment by sfan5** [comment]
While other improvements often end up with "this should be optional or it can't be accepted", I don't think that's needed for this change and it can be merged as-is.

**Comment by Zughy** [comment]
Any news? I'm not used to "what happens after a new release", if features must wait a while before getting considered or idk; just, I saw new core devs PRs popping up so I felt like asking

**Comment by LoneWolfHT** [comment]
As soon as a release is over from what I'm aware it's time to get features in again

**Comment by appgurueu** [comment]
Minetest uses font shadow to increase legibility. I notice font shadow being turned off in your screenshots?

**Comment by LoneWolfHT** [comment]
> Minetest uses font shadow to increase legibility. I notice font shadow being turned off in your screenshots?

See the OP and some of the comments

**Comment by appgurueu** [comment]
I wonder how this looks with font shadow enabled.

**Comment by Zughy** [comment]
Any news about it? One approval to go and it requires very little testing :(

**Comment by luk3yx** [comment]
Is it possible to turn off this background for empty/transparent nametags?

**Comment by SmallJoker** [comment]
Here's a comparison of the current shadow settings.
Left: `font_shadow = 0`, right `font_shadow = 1`
grafik

In the worst case scenario it is a bit difficult to read the player name, but still far better than without shadow.
The grey box behind text surely helps but at least to me it's not appealing visually. Semi-grey boxes are neutral, but also dull. If you'd put the green loading bar on startup behind names (with font shadows on) it might already look better.

**Comment by Zughy** [comment]
Bump

Also, I think the bar should be neutral. If even, it'd be cool if people could customise it, but this is for another PR.

Use-cases:
* better distinguish admins/VIPs/whoever in a server
* unlock cool backgrounds
* probably other things

**Comment by sfan5** [comment]
I don't get what you mean by "neutral by default" but customizing the tint of the bar sounds like a good future feature.

File: src/client/camera.cpp [hunk 1]
```diff
@@ -706,19 +707,20 @@ void Camera::drawNametags()
...
-			v2u32 screensize = RenderingEngine::get_video_driver()->getScreenSize();
 			v2s32 screen_pos;
 			screen_pos.X = screensize.X *
 				(0.5 * transformed_pos[0] * zDiv + 0.5) - textsize.Width / 2;
 			screen_pos.Y = screensize.Y *
 				(0.5 - transformed_pos[1] * zDiv * 0.5) - textsize.Height / 2;
 			core::rect<s32> size(0, 0, textsize.Width, textsize.Height);
-			g_fontengine->getFont()->draw(
+			core::rect<s32> bg_size(-2, 0, textsize.Width+2, textsize.Height);
+			driver->draw2DRectangle(video::SColor(50, 0, 0, 0),
```
Comment (appgurueu) [inline src/client/camera.cpp:721]: The color should not be hardcoded like this: What if a nametag happens to be grey? Will it be legible?

**Comment by Wuzzy2** [comment]
IIRC, nametag opacity is already settable, using the nametag color. You just have to do it explicitly in the game for all nametags.

**Comment by Wuzzy2** [comment]
Oops. Please forgive me! I have completely misunderstood the purpose of this. I take back my complaint 100%.
I support this now.

File: src/client/camera.cpp (hunk 1 above)
Comment (rubenwardy) [inline src/client/camera.cpp:721]: The background should also probably change color depending on the lightness.

Maybe something like this? Note that this is psuedocode, I can't remember the exact type and member names for SColor.

Source of weights: https://stackoverflow.com/questions/3942878/how-to-decide-font-color-in-white-or-black-depending-on-background-color

```c++
core::SColor textColor = nametag->nametag_color;
bool darkBackground = (textColor.red*0.299 + textColor.green*0.587 + textColor.blue*0.114) > 186;
const core::SColor backgroundColor = darkBackground ? video::SColor(50, 0, 0, 0) : video::SColor(50, 255, 255, 255);

**Comment by rubenwardy** [comment]
The test example shows a few examples where the light background is incorrectly chosen :P

**Comment by rubenwardy** [comment]
The threshold for the light/dark background was taken from the simpler method here: https://stackoverflow.com/questions/3942878/how-to-decide-font-color-in-white-or-black-depending-on-background-color

There's a more complex method of choosing the color using W3C recommendations that could give a better result. Alternatively, the threshold could be decreased so that more backgrounds are dark

**Comment by Zughy** [comment]
I actually prefer my version: nametag colours are more limited, but the eye knows exactly that dark gray box = ObjectRef name. On the contrary, with this one, even if it expands the range of nametag colours (some are still unreadable like the `4442` in the second screen), it doesn't create a standard. Also, in this way I have the impression some are more important than others, my brain feel kind of lost because it doesn't know where to focus. Personally :+1: for the lighter gray and :-1: for the changing background (also, less code). I'd rather stick with this solution https://github.com/minetest/minetest/pull/10152#issuecomment-691743011 , exposing it to the API

**Comment by rubenwardy** [comment]
A field can be added later to set the background, changing it from the default - it's not incompatible with the attempt to switch between white and black.

You could try reducing the threshold to make more black, say around 150, but I'm not too stuck with the white/black thing anyway. Feel free to revert it. Just please keep the devtest entity :)

**Comment by appgurueu** [comment]
Exposing it to the API is fine, but if the background color is not set, it should default to a background color ensuring legibility, for compatibility reasons.

File: src/client/camera.cpp [hunk 2]
```diff
@@ -706,21 +707,29 @@ void Camera::drawNametags()
...
 				(0.5 - transformed_pos[1] * zDiv * 0.5) - textsize.Height / 2;
 			core::rect<s32> size(0, 0, textsize.Width, textsize.Height);
-			g_fontengine->getFont()->draw(
+			core::rect<s32> bg_size(-2, 0, textsize.Width+2, textsize.Height);
+
+			video::SColor textColor = nametag->nametag_color;
+
+			bool darkBackground = textColor.getLuminance() > 186;
+			video::SColor backgroundColor = darkBackground
+					? video::SColor(50, 50, 50, 50)
```
Comment (appgurueu) [inline src/client/camera.cpp:726]: Alpha of 50?

Comment (rubenwardy) [reply src/client/camera.cpp:726]: Why not? That's out of 255, btw

**Comment by Zughy** [comment]
@rubenwardy I've tried your version and mine, did a couple of screens, compared them: definitely changed my mind, if you want to approve it just go for it :D and thanks for the help, really appreciated.

There is a bunch of colours that is not *that* readable, but overall it's a major improvement. I think is a step into the right direction, I'd avoid stalling here on details which can be added down the road

**Review by rubenwardy — APPROVED** [review]

**Comment by GenshinMT** [comment]
would be nice if modders are able to toggle the semi-transparent backgrounds

**Comment by Desour** [comment]
Can there please be a client-side setting for this?
Imho background boxes look a bit ugly.
(Client-side setting because it's subjective.)

**Comment by Zughy** [comment]
According to comments here and things said on IRC, I'd like to recap and propose a new approach:
* put a text outline instead of the semi-transparent box and remove the font shadow. Ie. [Among Us](https://i.ytimg.com/vi/rhHn2bjR1Ns/maxresdefault.jpg)
* implement somewhere down the road a server toggleable background (per player too) via API to do things like [this](https://gamefabrique.com/storage/screenshots/pc/maplestory-2-07.png) (Maple Story 2). This will be disabled by default 

Now, a problem could be the colour of the outline: as you can see in the first screen, a dark colour makes the text kind of unreadable, so the solution adopted by rubenwardy for the box colour might be applied here too

What do you think about it?

**Comment by rubenwardy** [comment]
Yeah, that makes sense to me

**Comment by Zughy** [comment]
One week passed, almost nobody said anything, I'm still more convinced of the outline and a core dev approves too. I don't have the time to do work on it as for now, but I'll try eventually. Closing this one

**Comment by Wuzzy2** [comment]
WTF?! Please don't kill a PR that has already been approved by 2 core devs.

Personally, I have no strong perference, but I like the background idea slightly more.

**Comment by LoneWolfHT** [comment]
These backgrounds are fine and can be improved later if necessary anyway

**Comment by Zughy** [comment]
Well, I guess there is some feedback now. Reopening `¯\_(ツ)_/¯`

(this could be modified when/if the outline PR gets done)

**Comment by paramat** [comment]
As long as a good implementation is found, i do not think making this optional is necessary.
Contrary to claims above, core devs have often opposed making something an option, i have many times.

> According to comments here and things said on IRC, I'd like to recap and propose a new approach:
put a text outline instead of the semi-transparent box and remove the font shadow

> the solution adopted by rubenwardy for the box colour might be applied here too

Cannot tell for sure until it is tried, but i think an outline is preferable to a background box, just as effective but less obtrusive.
But i do not object to this box implementation being merged for now.

**Comment by appgurueu** [comment]
The box implementation is inconsistent: What about waypoints, chat and other HUD text elements?

**Comment by Zughy** [comment]
> Cannot tell for sure until it is tried
A couple of weeks ago I tried understanding whether an outline was doable through CGUITT font or not, but I had no luck. That's probably me not being skilled enough though

> What about waypoints, chat and other HUD text elements?
Think about this as a testing ground and - possibly - as a transition. Also because chat had already been discussed (#10372 )

**Comment by SmallJoker** [comment]
The outline could be done using the current shadow effect, but repeated four times around the text using the offsets x±1 and y±1. However, doing that as an emboss effect in a shader would be way faster and probably look better than manual calculating.

**Comment by Zughy** [comment]
Bump. This only needs to be merged

**Comment by runsy** [comment]
Regression. Please, make the background disabled by default or optional instead.

**Comment by Wuzzy2** [comment]
If alpha is set to 0, the background disappears. I tested this.

**Comment by appgurueu** [comment]
> If alpha is set to 0, the background disappears. I tested this.

Yes, but this requires servers to upgrade to mods which actually set alpha to zero.

I still consider this as **breaking backwards-compatibility**, as a space-only nametag `" "` used to work, as well as an empty one `""` (*which has to be set using `set_nametag_attributes` and not `set_properties`, however!*). Both now draw a small empty, annoying box.

Additionally, consistent and more subtle display of nametags is now unachievable - every other text element, including waypoints, does not have a background. I agree with @runsy that this should definitely be configurable.

File: src/client/camera.cpp [hunk 3]
```diff
@@ -690,10 +690,11 @@ void Camera::drawNametags()
...
-	for (std::list<Nametag *>::const_iterator
-			i = m_nametags.begin();
-			i != m_nametags.end(); ++i) {
-		Nametag *nametag = *i;
+	gui::IGUIFont *font = g_fontengine->getFont();
+	video::IVideoDriver *driver = RenderingEngine::get_video_driver();
+	v2u32 screensize = driver->getScreenSize();
+
+	for (const Nametag *nametag : m_nametags) {
 		if (nametag->nametag_color.getAlpha() == 0) {
```
Comment (appgurueu) [inline src/client/camera.cpp:698]: this also needs ` && nametag->nametag_text != ""` for backwards compatibility not to be broken

**Comment by sfan5** [comment]
Note that formally backwards compatibility is only broken if the non-working behaviour was documented.

The old docs make no explicit mention of this.
Question here is is it reasonable to assume that rendering an empty or whitespace nametag would not render anything?

**Comment by LoneWolfHT** [comment]
In ~0.4.17 you could make nametags invisible by setting their alpha, but older clients didn't support that so the nametags would just be black. The way to fix that was to set the nametag to whitespace

**Comment by Wuzzy2** [comment]
> Question here is is it reasonable to assume that rendering an empty or whitespace nametag would not render anything?

Empty: Yes.
Whitespace-only: Maybe.

**Comment by runsy** [comment]
A blank space is a char:
https://emptycharacter.com/

**Pull request merged**